### PR TITLE
Add Log.warn_once

### PR DESCRIPTION
**Pull request by jkeiser**

### CODE REVIEW

File: lib/mixlib/log.rb [hunk 1]
```diff
@@ -30,7 +31,7 @@ module Log
 
 
     def reset!
-      @logger, @loggers = nil, nil
+      @logger, @loggers, @warned_from = nil, nil
```
Comment (coderanger) [inline lib/mixlib/log.rb]: Probably means this to be `= nil, nil, nil`. Related, if you put these on their own lines these bugs would be more obvious :)

**Comment by coderanger** [comment]
:+1: Was just asking for this over on another ticket :)

Only question is that right now this limits to once per call. Would it maybe be better to do it once per distinct message?

**Comment by jkeiser** [comment]
@coderanger I thought about that. The specific case I'm using it for *now* is really once per caller, and I think the others I would use it on might be that way.  But I can imagine cases that aren't. Got an idea of how we could name the methods to make the distinction clear?

One possibility
`warn_once()` - warns once per caller.
`warn_once_per_unique()` - warns once per unique warning text.

If those don't sound terrible, I won't feel bad checking this one in, and adding that in a separate PR.

**Comment by coderanger** [comment]
No, we should pick one and go with it for API cleanliness. "once per message" seems like the superset case since you can always make a message be the same in all cases to only show it once ever. That seems like less of a footgun, it forces things to be clear that, for example, with the freeze warning you would see it once ever, instead of once per LWRP or whatnot.

**Comment by jkeiser** [comment]
Sure. In my case though, I *want* something that will only ever report the first one ever. The rest will show up in debug, so people can get them if they want; everyone else has just a little less noise.

**Comment by jkeiser** [comment]
And I can't make the messages the same because I need to show which property of which resource (in this case) is being declared with a non-frozen array default.

**Comment by coderanger** [comment]
That seems like unexpected behavior for most people. There is nothing special about the first one specifically, whereas suppressing actual duplicates is a clearer semantic.

**Comment by jkeiser** [comment]
I'm OK with making the method you are suggesting, but then I still need the method that does what I'm trying to do :)

**Comment by coderanger** [comment]
Maybe as a way to combine them, have a `:once_key` kwarg that is what is used for the once-ness tracking. Can default to the message string itself but for the freeze warning it could be set to any static string to get a "once overall" semantic.

**Comment by jkeiser** [comment]
Actually, it might work best if we had a `dedup` keyword arg that defaulted to `true` for Chef::Log.warn. It seems like every single warning I've looked at, that's what I really wanted. Then `warn_once` would do what it does in this patch, and everybody gets the thing they want.

**Comment by coderanger** [comment]
Ehh, I agree that would be better but it also seems like a hard thing to document that the warn level happens to be special in that way. Making it a different method makes the different-ness more understandable.

**Comment by lamont-granquist** [comment]
We also need a Chef::Config setting to unset this and un-dedup all the warns for those of us who think the existing behavior is correct =)

**Comment by coderanger** [comment]
@lamont-granquist Either way it gets implemented, you would see the duplicates if you turned up to `:debug`. Can't tell if you are joking but a config option wouldn't be appropriate for this.

**Comment by lamont-granquist** [comment]
Not joking.  I don't want to look at debug logs, and want to see all the duplicated warns and get the existing behavior.

**Comment by coderanger** [comment]
Then don't use `warn_once` for those?

**Comment by lamont-granquist** [comment]
No, I want to turn all the `warn_once` calls that jkeiser writes into `warn` calls when I run chef-client on my servers.  That way if there's 1,000 of them kicked out in a chef-client run I don't get only the first one and either miss it or think its no big deal, and i don't have to go sifting through debug logs to see that there's really 1,000 of them.

**Comment by lamont-granquist** [comment]
And if we do this we probably need a count of warns (and/or count of suppressed warns) added to the end of the chef-client run.

**Comment by coderanger** [comment]
My point is that if it is something where seeing it more than once is important, don't use `warn_once`. If you are saying you think people will ignore warnings unless there is an avalanche of them, then we need to re-educate users about what a warning means.

**Comment by coderanger** [comment]
Any `WARN` line in Chef's output should be a big deal. We've just been terrible about output cleanliness and trained users to ignore warnings as non-actionable.

**Comment by lamont-granquist** [comment]
And my point is that as a user, I want to not suppress any warns.  It isn't about when I'm developing and writing the warns.  And yes, that's about treating the warns seriously, I don't want them suppressed, and I tend to see this feature as going the wrong direction in terms of treating them seriously.  If you suppress all-but-one then it seems like less of a deal to the end user and the output seems cleaner, when its not, so it allows them to continue inaction.

But, i'm okay with adding this feature and even adding it by default, but I simply want a way to turn it off.

**Comment by coderanger** [comment]
If we think users don't want warnings de-duplicated then we shouldn't add this feature.

**Comment by lamont-granquist** [comment]
So people seem to want it.  And I can see the argument about it, even though I disagree.  Other tools like rspec have this feature.  They also have the ability, however, to turn off the deduplication of warnings.  And while I don't get the argument in favor of deduplication, I really don't understand the argument about making it configurable.  It seems an obvious switch to add.

**Comment by coderanger** [comment]
Because we have a metric ass-load of niche configuration options, so many that expecting a user to be able to easily discover this new one would be tricky (would have to make the end-of-run output specifically mention it?) and even with that options bloat is a problem. I think we need a more concrete story for the option than "I disagree".

**Comment by coderanger** [comment]
Assuming we add this and the default is to de-dup, who is the user than wants to change that, how do they figure out they want to change it, and then how do they learn what action to take? And how do we do all that without making life harder for the (presumably or we shouldn't add the feature) majority case that doesn't want duplicate warnings.

**Comment by danielsdeleo** [comment]
My preference is to collect them all, then report at the end of the run:

```
You have issue X. The reason for it is Y and the fix is Z.

this happened N times in these specific places:
- A
- B
- C
```

That said, when I'm actively fixing a problem, I prefer setting config to turn warnings into errors and then just running the code until it stops failing, and we already have that.

**Comment by coderanger** [comment]
@danielsdeleo That sounds like a much more constructive user story :+1:

**Comment by thommay** [comment]
Closing…

**Pull request closed without merging**